File: frontend/src/utils/pixelCrdt.ts
/** 稀疏像素 CRDT：key = "x,y" → label。与 backend pixel_crdt 对齐。 */

export const PIXEL_CAP = 200_000

export function pixelKey(x: number, y: number): string {
  return `${Math.round(x)},${Math.round(y)}`
}

export function parsePixelKey(key: string): [number, number] {
  const [a, b] = key.split(',')
  return [Number(a), Number(b)]
}

export function paintDisk(
  pixels: Record<string, string>,
  cx: number,
  cy: number,
  radius: number,
  label: string,
  erase = false,
  cap = PIXEL_CAP,
): Record<string, string> {
  const next = { ...pixels }
  const r = Math.max(1, Math.round(radius))
  const r2 = r * r
  const x0 = Math.round(cx)
  const y0 = Math.round(cy)
  for (let dy = -r; dy <= r; dy++) {
    for (let dx = -r; dx <= r; dx++) {
      if (dx * dx + dy * dy > r2) continue
      const k = pixelKey(x0 + dx, y0 + dy)
      if (erase) delete next[k]
      else if (next[k] !== undefined || Object.keys(next).length < cap) next[k] = label
    }
  }
  return next
}

File: backend/app/services/pixel_crdt.py
"""稀疏像素 / 点标签 CRDT 辅助（与前端 lib/pixelCrdt.ts 对齐）。"""

from __future__ import annotations

from typing import Dict, Iterable, Optional, Tuple

PIXEL_CAP = 200_000


def pixel_key(x: int, y: int) -> str:
    return f"{int(x)},{int(y)}"


def parse_pixel_key(key: str) -> Tuple[int, int]:
    a, b = str(key).split(",", 1)
    return int(a), int(b)


def paint_disk(
    pixels: Dict[str, str],
    cx: int,
    cy: int,
    radius: int,
    label: str,
    *,
    erase: bool = False,
    cap: int = PIXEL_CAP,
) -> Dict[str, str]:
    r = max(1, int(radius))
    r2 = r * r
    for dy in range(-r, r + 1):
        for dx in range(-r, r + 1):
            if dx * dx + dy * dy > r2:
                continue
            k = pixel_key(cx + dx, cy + dy)
            if erase:
                pixels.pop(k, None)
            elif k in pixels or len(pixels) < cap:
                pixels[k] = label
    return pixels


def merge_point_labels(
    current: Dict[str, str],
    patch: Dict[str, Optional[str]],
) -> Dict[str, str]:
    out = dict(current)
    for k, v in patch.items():
        if v is None or v == "":
            out.pop(str(k), None)
        else:
            out[str(k)] = str(v)
    return out


def object_keys(kind: str, ids: Iterable[str]) -> list[str]:
    prefix = "a:" if kind == "ann" else "b:"
    return [f"{prefix}{i}" for i in ids]
